### PR TITLE
feat: adjust thresholds for certificate health

### LINKED ISSUE #101: Adjust thresholds for certificates health expiry date
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

With VCF 4.4 SOS reports RED for certificates if it expires in < 1 day. else its GREEN.

The thresholds have been updated in VCF 4.5 SOS Utility for Certificates Health data to -

```
<15 days -- RED
15-30 days -- YELLOW
>30 days -- GREEN
```

We should keep the thresholds same for 4.4 and 4.5.

This needs to be handled in powershell code so that the alert column is set appropriately in the report. 
This doesn't impact HRM solution as we are getting SOS data directly from python code using SDDC Manager API. 

### Use Case(s)

Currently thresholds are <1 - RED and >1 GREEN. These were set based on values in VCF 4.4

We should be consistent and should use 15 (RED) < days (YELLOW) < 30 (GREEN) thresholds across, irrespective of what SOS reports. 

### Potential Configuration

Thresholds for Certificates health should be -

```
<15 days -- RED
15-30 days -- YELLOW
>30 days -- GREEN
```

### References

_No response_

### PR DESCRIPTION
**Pull request by bhumitra**
In order to have a good experience with our community, we recommend that you read the [contributing guidelines](https://github.com/vmware/powershell-module-for-cloud-foundation-reorting/blob/main/CONTRIBUTING_DCO.md) for making a pull request.

**Summary of Pull Request**

- Adjust thresholds for certificate health expiry
- Added ESXi hosts data back into the report
- Added certificate expires in days column to report

**Type of Pull Request**

- [ ] This is a bug fix.
- [x] This is an enhancement or feature.
- [ ] This is a code style / formatting update.
- [ ] This is a documentation update.
- [ ] This is a refactoring update.
- [ ] This is a chore update
- [ ] This is something else.
      Please describe:

**Related to Existing Issues**

Closes #101 

**Test and Documentation Coverage**

- [x] Tests have been completed (for bug fixes / features).
- [ ] Documentation has been added / updated (for bug fixes / features).

**Breaking Changes?**

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

### CODE REVIEW

**Comment by bhumitra** [comment]
image

This is what it looks like now.

**Review by tenthirtyam — APPROVED** [review]
LGTM! 🚀

**Comment by tenthirtyam** [comment]
Approve but will merge soon.

**Review by tenthirtyam — DISMISSED** [review]
Error observed on Windows:

```powershell
PS F:\VMware.CloudFoundation.Reporting> Import-Module .\VMware.CloudFoundation.Reporting.psd1
PS F:\VMware.CloudFoundation.Reporting> Publish-CertificateHealth -json F:\Reporting\HealthReports\sfo-w01-all-health-results.json
 [03-28-2023_14:17:05]   Error at Script Line 926
 [03-28-2023_14:17:05]   Relevant Command: $expiry = [DateTime]::ParseExact($element.title[2], "MMM dd HH:mm:ss yyyy 'GMT'", $null).ToUniversalTime()
 [03-28-2023_14:17:05]   Error Message: Exception calling "ParseExact" with "3" argument(s): "String was not recognized as a valid DateTime."
PS F:\VMware.CloudFoundation.Reporting> Publish-CertificateHealth -json F:\Reporting\HealthReports\sfo-vcf01-all-health-results.json
 [03-28-2023_14:17:22]   Error at Script Line 926
 [03-28-2023_14:17:22]   Relevant Command: $expiry = [DateTime]::ParseExact($element.title[2], "MMM dd HH:mm:ss yyyy 'GMT'", $null).ToUniversalTime()
 [03-28-2023_14:17:22]   Error Message: Exception calling "ParseExact" with "3" argument(s): "String was not recognized as a valid DateTime."
```

**Comment by tenthirtyam** [comment]
Marked as draft whilst awaiting changes.

File: VMware.CloudFoundation.Reporting.psm1 [hunk 1]
```diff
@@ -916,13 +916,26 @@ Function Publish-CertificateHealth {
...
-            $jsonInputData.PSObject.Properties.Remove('ESXI')
             foreach ($component in $jsonInputData.PsObject.Properties.Value) {
                 foreach ($element in $component.PsObject.Properties.Value) {
                     $elementObject = New-Object -TypeName psobject
                     $elementObject | Add-Member -NotePropertyName 'Component' -NotePropertyValue ($element.area -Split (':'))[0].Trim()
                     $elementObject | Add-Member -NotePropertyName 'Resource' -NotePropertyValue ($element.area -Split (':'))[-1].Trim()
-                    $elementObject | Add-Member -NotePropertyName 'Alert' -NotePropertyValue $element.alert
+                    if ($element.title -and $element.title -notcontains "-") {
+                        $current = [DateTime]::Now
+                        $expiry = [DateTime]::ParseExact($element.title[2], "MMM dd HH:mm:ss yyyy 'GMT'", $null).ToUniversalTime()
```
Comment (bhumitra) [inline VMware.CloudFoundation.Reporting.psm1:926]:
```suggestion
                        $expiry = [DateTime]::ParseExact(($element.title[2] -replace '\s+', ' '), "MMM d HH:mm:ss yyyy 'GMT'", $null).ToUniversalTime()
```

**Comment by tenthirtyam** [comment]
Looks good after testing with the update but I do see an oddity where it’s reporting a -25 but an 8 day expiration. You can test with the JSON I provided over Slack. I suspect this is a data issue from SoS?

```powershell
PS F:\VMware.CloudFoundation.Reporting> Publish-CertificateHealth -json F:\Reporting\HealthReports\sfo-vcf01-all-health-results.json -failureOnly


Component         : vRLI
Resource          : sfo-vrli01a.sfo.rainpole.io
Expires In (Days) : -25
Alert             : RED
Message           : Certificate validation is failed for vRLI:sfo-vrli01a.sfo.rainpole.io. If Certificate End date is 15 days or less, SoS will show result as failed.Certificate expires in 8
                    day(s).
```

**Comment by tenthirtyam** [comment]
I’ll submit an update for the build and changelog before merge.

**Review by tenthirtyam — APPROVED** [review]
LGTM! 🚀

**Pull request merged**